doc/namespace.md

```diff
@@ -0,0 +1,417 @@
+
+#### Size of RClass and rb_classext_t
+
+Namespace requires to move some fields from RClass to `rb_classext_t`, then the size of RClass and `rb_classext_t` is now larger than `4 * RVALUE_SIZE`. It's against the expectation of [Variable Width Allocation](https://rubykaigi.org/2021-takeout/presentations/peterzhu2118.html).
+
+Now the `STATIC_ASSERT` to check the size is commented-out. (See "Minimize the size of RClass and rb_classext_t" in "Discussion" section below)
+
+#### ISeq inline method/constant cache
+
+As described above in "Namespace scopes", an ".rb" file runs in a namespace. So method/constant resolution will be done in a namespace consistently.
+
+That means ISeq inline caches work well even with namespaces. Otherwise, it's a bug.
+
+#### Method call global cache (gccct)
+
+`rb_funcall()` C function refers to the global cc cache table (gccct), and the cache key is calculated with the current namespace.
+
+So, `rb_funcall()` calls have a performance penalty when namespace is enabled.
+
+#### Current namespace and loading namespace
+
+The current namespace is the namespace that the executing code is in. `Namespace.current` returns the current namespace object.
+
+The loading namespace is an internally managed namespace to determine the namespace to load newly required/loaded files. For example, `ns` is the loading namespace when `ns.require("foo")` is called.
+
+## Discussions
+
+#### Namespace#inspect
+
+Currently, `Namespace#inspect` returns values like `"#<Namespace:0x00000001083a5660>"`. This results in the very redundant and poorly visible classpath outside the namespace.
+
+```ruby
+# foo.rb
+class C; end
+
+# main.rb
+ns = Namespace.new
+ns.require('foo')
+
+p ns::C # "#<Namespace:0x00000001083a5660>::C"
+```
+
+And currently, if a namespace is assigned to a constant `NS1`, the classpath output will be `NS1::C`. But the namespace object can be brought to another namespace and the constant `NS1` in the namespace is something different. So the constant-based classpath for namespace is not safe basically.
+
+So we should find a better format to show namespaces. Options are:
+
+* `NS1::C` (only when this namespace is created and assigned to NS1 in the current namespace)
+* `#<Namespace:user:1083a5660>::C` (with namespace type and without preceding 0)
+* or something else
+
+#### Namespace#eval
+
+Testing namespace features needs to create files to be loaded in namespaces. It's not easy nor casual.
+
+If `Namespace` class has an instance method `#eval` to evaluate code in the namespace, it can be helpful.
+
+#### More builtin methods written in Ruby
+
+If namespace is enabled by default, builtin methods can be written in Ruby because it can't be overridden by users' monkey patches. Builtin Ruby methods can be JIT-ed, and it could bring performance reward.
```

Comment:
> it can't be overridden by users' monkey patches

It would be good to expand this to explain what it would solve concretely with examples.
In https://bugs.ruby-lang.org/issues/21311#note-36 I wrote that method monkey patches are usually dealt with Primitives.
Maybe Namespace would make this nicer/less necessary?
Namespace I guess could make `Primitive.is_a?(obj, String)` safer in the case `String` would be re-assigned by some code (that's very unlikely though).